src/syscheckd/db/src/db.cpp

```diff
@@ -20,17 +20,19 @@ extern "C" {
 /**
  * @brief Create the statement string to create the dbsync schema.
  *
- * @return std::string Contains the dbsync's schema for FIM db.
+ * @return char* Contains the dbsync's schema for FIM db.
  */
-const char * CreateStatement()
+static char* CreateStatement()
```

Comment:
We could change this with tempered factory pattern. I will pass a example with this change.
```suggestion
diff --git a/src/syscheckd/db/CMakeLists.txt b/src/syscheckd/db/CMakeLists.txt
index 6634b8152c..10bd0a31b9 100644
--- a/src/syscheckd/db/CMakeLists.txt
+++ b/src/syscheckd/db/CMakeLists.txt
@@ -71,6 +71,9 @@ link_directories(${SRC_FOLDER}/external/openssl/)
 
 if(CMAKE_SYSTEM_NAME STREQUAL "Windows")
   add_definitions(-DWIN32=1 -D_WIN32_WINNT=0x600 -DWIN_EXPORT)
+  add_definitions(-DOS_TYPE=OSType::WINDOWS)
+else()
+  add_definitions(-DOS_TYPE=OSType::LINUX)
 endif(CMAKE_SYSTEM_NAME STREQUAL "Windows")
 
 if(CMAKE_SYSTEM_NAME STREQUAL "Windows")
diff --git a/src/syscheckd/db/include/fimCommonDefs.h b/src/syscheckd/db/include/fimCommonDefs.h
index e997905e78..14cebfc413 100644
--- a/src/syscheckd/db/include/fimCommonDefs.h
+++ b/src/syscheckd/db/include/fimCommonDefs.h
@@ -20,4 +20,9 @@ typedef enum FIMDBErrorCodes
 typedef void((*fim_sync_callback_t)(const char *tag, const char* buffer));
 typedef void((*logging_callback_t)(const modules_log_level_t level, const char* log));
 
+enum OSType
+{
+    LINUX,
+    WINDOWS
+};
 #endif // DB_STATEMENT_H
diff --git a/src/syscheckd/db/include/fimDBHelper.hpp b/src/syscheckd/db/include/fimDBHelper.hpp
index 9ddc8c13dd..1e27d8c192 100644
--- a/src/syscheckd/db/include/fimDBHelper.hpp
+++ b/src/syscheckd/db/include/fimDBHelper.hpp
@@ -13,6 +13,7 @@
 #define _FIMDBHELPER_HPP
 #include "fimDB.hpp"
 #include "dbItem.hpp"
+#include "fimCommonDefs.h"
 
 namespace FIMDBHelper
 {
@@ -220,6 +221,42 @@ namespace FIMDBHelper
 
         return query;
     }
+
+    // Exception template
+    template <OSType type>
+    class FactoryCreateStatements final
+    {
+        public:
+            static std::string CreateStatement()
+            {
+                throw std::runtime_error
+                {
+                    "There was a problem when db statement generation"
+                };
+            }
+    };
+    // Standard template to extract package information in fully compatible Linux systems
+    template <>
+    class FactoryCreateStatements<OSType::WINDOWS> final
+    {
+        public:
+            static std::string CreateStatement()
+            {
+                return std::string(CREATE_FILE_DB_STATEMENT) + std::string(CREATE_REGISTRY_KEY_DB_STATEMENT) + std::string(CREATE_REGISTRY_VALUE_DB_STATEMENT);
+            }
+    };
+
+    // Template to extract package information in partially incompatible Linux systems
+    template <>
+    class FactoryCreateStatements<OSType::LINUX> final
+    {
+        public:
+            static std::string CreateStatement()
+            {
+                return CREATE_FILE_DB_STATEMENT;
+            }
+    };
+
 }
 
 #endif //_FIMDBHELPER_H
diff --git a/src/syscheckd/db/src/db.cpp b/src/syscheckd/db/src/db.cpp
index 73fa7172bd..b3d08dd4f2 100644
--- a/src/syscheckd/db/src/db.cpp
+++ b/src/syscheckd/db/src/db.cpp
@@ -22,16 +22,16 @@ extern "C" {
  *
  * @return std::string Contains the dbsync's schema for FIM db.
  */
-static char *CreateStatement()
-{
-    std::string ret = CREATE_FILE_DB_STATEMENT;
-#ifdef WIN32
-    ret += CREATE_REGISTRY_KEY_DB_STATEMENT;
-    ret += CREATE_REGISTRY_VALUE_DB_STATEMENT;
-#endif
+// static char *CreateStatement()
+// {
+//     std::string ret = CREATE_FILE_DB_STATEMENT;
+// #ifdef WIN32
+//     ret += CREATE_REGISTRY_KEY_DB_STATEMENT;
+//     ret += CREATE_REGISTRY_VALUE_DB_STATEMENT;
+// #endif
 
-    return strdup(ret.c_str());
-}
+//     return strdup(ret.c_str());
+// }
 
 #ifndef WIN32
 void fim_db_init(int storage,
@@ -51,8 +51,11 @@ void fim_db_init(int storage,
     try
     {
         auto path = (storage == FIM_DB_MEMORY) ? FIM_DB_MEMORY_PATH : FIM_DB_DISK_PATH;
-        char *createQuery =  CreateStatement();
-        auto dbsyncHandler = std::make_shared<DBSync>(HostType::AGENT, DbEngineType::SQLITE3, path, createQuery);
+        //char *createQuery =  CreateStatement();
+        auto dbsyncHandler = std::make_shared<DBSync>(HostType::AGENT,
+                                                      DbEngineType::SQLITE3,
+                                                      path,
+                                                      FIMDBHelper::FactoryCreateStatements<OS_TYPE>::CreateStatement());
         auto rsyncHandler = std::make_shared<RemoteSync>();
 
 #ifndef WIN32
@@ -61,7 +64,6 @@ void fim_db_init(int storage,
         FIMDBHelper::initDB<FIMDB>(sync_interval, file_limit, value_limit, sync_callback, log_callback, dbsyncHandler,
                                    rsyncHandler);
 #endif
-        free(createQuery);
     }
     catch (const DbSync::dbsync_error& ex)
     {
```